microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/core/PropertiesManagerForJava.java

```diff
@@ -370,6 +371,42 @@ public Hover hover(MicroProfileJavaHoverParams params, IJDTUtils utils, IProgres
 		return hovers.get(0);
 	}
 
+	/**
+	 * Returns the cursor context for the given file and cursor position, or NONE_OF_THE_ABOVE if none of the cursor contexts are applicable.
+	 *
+	 * @param params the completion params that provide the file and cursor position to get the context for
+	 * @param utils the jdt utils
+	 * @param monitor the progress monitor
+	 * @return the cursor context for the given file and cursor position, or null if none of the cursor contexts are applicable
+	 * @throws JavaModelException when the buffer for the file cannot be accessed or the Java model cannot be accessed
+	 */
+	public JavaCursorContext snippetCursorContext(MicroProfileJavaCompletionParams params, IJDTUtils utils,
+			IProgressMonitor monitor) throws JavaModelException {
+		String uri = params.getUri();
+		ITypeRoot typeRoot = resolveTypeRoot(uri, utils, monitor);
+		if (typeRoot == null || typeRoot.findPrimaryType() == null) {
+			return JavaCursorContext.EMPTY_FILE;
+		}
+
+		Position completionPosition = params.getPosition();
+		int completionOffset = utils.toOffset(typeRoot.getBuffer(), completionPosition.getLine(),
+				completionPosition.getCharacter());
+
+		IJavaElement element = typeRoot.getElementAt(completionOffset);

```

Comment:
I think you should use AST CompilationUnit instead of DOM CompilationUnit to be more precise. To get AST CompilationUnit, see https://github.com/eclipse/lsp4mp/blob/caf91c68e7f2fe27072f1f947e3c5476f3289d8f/microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/core/java/AbtractJavaContext.java#L98 and after that you get the covered node https://github.com/eclipse/lsp4mp/blob/f5571c37f71799acd5c0a0a14b43b735bee83313/microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/core/java/codeaction/JavaCodeActionContext.java#L94